services/venice-server/src/main/java/com/linkedin/venice/listener/ReadQuotaEnforcementHandler.java

```diff
@@ -192,26 +207,30 @@ public void channelRead0(ChannelHandlerContext ctx, RouterRequest request) {
       return;
     }
 
-    if (checkInitAndQuotaEnabled(ctx, request, store, false)) {
+    if (checkInitAndQuotaEnabledToSkipQuotaEnforcement(ctx, request, store, false)) {
       return;
     }
 
     int rcu = getRcu(request); // read capacity units
 
     /**
-     * First check store bucket for capacity; don't throttle retried request at store version level
+     * First check store bucket for capacity don't throttle retried request at store version level
      */
     TokenBucket tokenBucket = storeVersionBuckets.get(request.getResourceName());
-    if (tokenBucket != null && !request.isRetryRequest()) {
-      if (!tokenBucket.tryConsume(rcu)) {
+    if (tokenBucket != null) {
+      if (!request.isRetryRequest() && !tokenBucket.tryConsume(rcu)
+          && handleTooManyRequests(ctx, request, null, store, rcu, false)) {
+        // Enforce store version quota for non-retry requests.
         // TODO: check if extra node capacity and can still process this request out of quota
-        if (handleTooManyRequests(ctx, request, null, store, rcu, false))
-          return;
+        return;
+      }
+    } else {
+      // If this happens it is probably due to a short-lived race condition where the resource is being accessed before
+      // the bucket is allocated. The request will be allowed based on node/server capacity so emit metrics accordingly.
+      stats.recordAllowedUnintentionally(storeName, rcu);
+      if (!noBucketStores.contains(request.getResourceName())) {
```

Comment:
may use `isEnforcingAndNoBucketStoreContainsResource` ?

Reply:
Will use `isNoBucketStoreContainsResource`, renamed the method since we removed the enforcing flag to avoid confusion.